### PR TITLE
DEPR: Adjust read excel behavior for xlrd >= 2.0

### LINKED ISSUE #38424: shift default excel read engine from xlrd to openpyxl
since v2.0.0, xlrd no longer supports excel files other than ".xls".  manually specify pd.read_excel engine (to openpyxl and etc) is a bit annoying (otherwise pandas would complain about the missing xlrd module).  is it possible to shift the default engine to sth more commonly used (eg. openpyxl )

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #38424
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Alternative to #38522. I've been testing this locally using both xlrd 1.2.0 and 2.0.1.

One test fails because we used to default to xlrd but now default to openpyxl, it's not clear to me if this test should be passing with openpyxl.

cc @cjw296, @jreback, @jorisvandenbossche

### CODE REVIEW

**Comment by rhshadrach** [comment]
Windows failures are related, will be investigating.

File: pandas/io/excel/_base.py
Comment (cjw296) [inline pandas/io/excel/_base.py:933]: Why provide degraded inference just because the source is a URL?
It appears this code goes out of its way to avoid using seek, whereas the ODS inference code that was there before, and xlrd which is being hit in many of the existing code paths already does seek without issue.

**Comment by cjw296** [comment]
I believe this would also replace #38456.

I find the desire to support xlrd 1.2 at a time when pandas users are already upgrading a package to be both surprising and disappointing, given the potential security issues and poor parsing experience associated with sticking with xlrd 1.2.

This PR has code that is more complex than #38424, even ignoring that which is in place to support the convoluted deprecation process it advocates for. I haven't seen code coverage metrics as part of the pandas PR process, but are you sure all the new conditional branches you've introduced are covered by sufficient tests?

In case it wasn't clear in  #38522, I believe the most robust approach in this area is to:
- obtain the stream we're going to end up with anyway
- peek the minimum number of bytes from it to do content-based inference
- `.seek(0)` to get the stream back to its initial state.

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py:933]: Previous ODS code read at most 84 bytes; current code needs the entire file. I'll do a partial revert here and utilize the previous ODS code, but I have reservations about downloading the entire file here. Would like to hear others' thoughts.

Comment (rhshadrach) [reply pandas/io/excel/_base.py:933]: Ah; I think I have falsely assumed we need to get the entire contents of the file. I think it should be possible to get BufferedIOBase/RawIOBase into the proper form for ZipFile without reading.

Comment (cjw296) [reply pandas/io/excel/_base.py:933]: I did put quite a lot of effort into #38424; the code in that PR is the way it is from, as carefully as I could, following through the various code paths and ensuring the behaviour was as simple and robust as it could be, in spite of less automated testing than I was expecting. It's tough to see these kind of comments which sort of imply that I hadn't thought any of this through...

Comment (rhshadrach) [reply pandas/io/excel/_base.py:933]: In no way shape or form did I have any intention of implying such a thing. As my comment above says, I was mistaken.

**Comment by rhshadrach** [comment]
Thanks for the comments @cjw296, I was able to simplify/improve a lot from them.

**Comment by cjw296** [comment]
@rhshadrach - I'm confused as to why you didn't just start with  #38424 and add the fallback logic you're advocating for.

**Comment by rhshadrach** [comment]
@cjw296 

> I'm confused as to why you didn't just start with #38424 and add the fallback logic you're advocating for.

I think you mean your PRs #38456/#38522. I had started this work 7 days ago, prior to the existence of them.

**Comment by rhshadrach** [comment]
2 linux tests failed to start, all other checks passed. Rerunning.

**Comment by rhshadrach** [comment]
Failure is related, though I don't understand it:

```
>               handle = open(handle, ioargs.mode)
E               FileNotFoundError: [Errno 2] 没有那个文件或目录: 'foo.xlsm'

[snip]

with pytest.raises(FileNotFoundError, match="No such file or directory"):
>           pd.read_excel(bad_file)
E           AssertionError: Regex pattern 'No such file or directory' does not match "[Errno 2] 没有那个文件或目录: 'foo.xlsm'".
```

**Review by jreback — APPROVED** [review]
lgtm. @jorisvandenbossche @simonjayhawkins

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py:1093]: Ah - yes, thanks. Updating now, should be quick.

**Comment by jorisvandenbossche** [comment]
Regardless of which PR was started first, it's *also* an option to update https://github.com/pandas-dev/pandas/pull/38522 to use the behaviour we want (fallback to xlrd if openpyxl is not installed, for xlrd < 2.0, see also https://github.com/pandas-dev/pandas/pull/38522#pullrequestreview-554024616, I think it should be fairly straightforward to add that to that PR)

**Comment by jreback** [comment]
> Regardless of which PR was started first, it's _also_ an option to update #38522 to use the behaviour we want (fallback to xlrd if openpyxl is not installed, for xlrd < 2.0, see also [#38522 (review)](https://github.com/pandas-dev/pandas/pull/38522#pullrequestreview-554024616), I think it should be fairly straightforward to add that to that PR)

ok sure. let's try to get this in today and cut the release. i view better inference as good but shouldn't hold this up.

**Comment by rhshadrach** [comment]
@jreback @jorisvandenbossche Is there room for improving the inference done here? As far as I can tell, the inference done here is that of #38522 but also handles bytes and urls (which I believe is causing failures in #38522).

**Comment by rhshadrach** [comment]
Two of the travis builds failed to start, the third one passed. Restarted manually.

File: pandas/io/excel/_base.py
Comment (jreback) [reply pandas/io/excel/_base.py:1094]: does this need updating?

**Review by jreback — CHANGES_REQUESTED** [review]
very minor comment. pls commetn / update

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py:1094]: xlrd_version is a LooseVersion here; LooseVersion comparison will handle `other` as being a string; but can replace with `LooseVersion("2")` if that's preferred.

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [inline pandas/io/excel/_base.py:129]: Can you add that it will raise an error in the future as well?

File: pandas/tests/io/excel/test_writers.py
Comment (rhshadrach) [reply pandas/tests/io/excel/test_writers.py]: Issue #38644 created and updated the test with the issue number.

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [inline pandas/io/excel/_base.py:1085]: So we should probably not silence the warning when `engine="xlrd"` is specified? Since in the future we want to disallow any reading of non-xls files with xlrd?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
great let's merge on greenish

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [inline pandas/io/excel/_base.py:1089]: What happens if xlrd>=2 is installed (and openpyxl not), then here no engine is assigned?

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [inline pandas/io/excel/_base.py:1099]: Can you add the recommendation of installing openpyxl?

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [inline pandas/io/excel/_base.py:1031]: Why is this removed here? I see it is moved to the top, but you can still end up here with a wrong/unknown engine at this point?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
> @jreback @jorisvandenbossche Is there room for improving the inference done here? As far as I can tell, the inference done here is that of #38522 but also handles bytes and urls (which I believe is causing failures in #38522).

Yeah, that's certainly fine for me as well. I am just not very familiar with the details, so can't really judge whether your version also covers the different edge cases that @cjw296 took care of in https://github.com/pandas-dev/pandas/pull/38522 (cfr https://github.com/pandas-dev/pandas/pull/38571#discussion_r546235310)

**Comment by jreback** [comment]
> > @jreback @jorisvandenbossche Is there room for improving the inference done here? As far as I can tell, the inference done here is that of #38522 but also handles bytes and urls (which I believe is causing failures in #38522).
> 
> Yeah, that's certainly fine for me as well. I am just not very familiar with the details, so can't really judge whether your version also covers the different edge cases that @cjw296 took care of in #38522 (cfr [#38571 (comment)](https://github.com/pandas-dev/pandas/pull/38571#discussion_r546235310))

sure, let's just merge this and can always do another update

**Comment by jorisvandenbossche** [comment]
@rhshadrach can you unping xlrd in some of the CI builds, so we ensure to actually test the case with xlrd == 2.0

**Comment by cjw296** [comment]
> sure, let's just merge this and can always do another update

I'd prefer it if you added the warnings and fallback in #38522 and use that. That code certainly works with bytes and urls, I remember fixing the tests. I stopped short of cleaning up the remaining test failures because it was unclear if the branch stood any chance of being merged, and I believe it's code is cleaner and simpler than the code in this PR.

**Comment by rhshadrach** [comment]
@jorisvandenbossche I have the changes made, but unsure of which CI builds to unpin; any guidance here? Or should I just pick a few?

**Comment by jreback** [comment]
> @jorisvandenbossche I have the changes made, but unsure of which CI builds to unpin; any guidance here? Or should I just pick a few?

```
jreback@lhs2:~/pandas-dev/ci$ grep xlrd deps/*
deps/azure-37-locale_slow.yaml:  - xlrd=1.2.0
deps/azure-37-minimum_versions.yaml:  - xlrd=1.2.0
deps/azure-37-slow.yaml:  - xlrd<2.0
deps/azure-38-locale.yaml:  - xlrd<2.0
deps/azure-38-slow.yaml:  - xlrd<2.0
deps/azure-macos-37.yaml:  - xlrd<2.0
deps/azure-windows-37.yaml:  - xlrd<2.0
deps/azure-windows-38.yaml:  - xlrd<2.0
deps/travis-37-cov.yaml:  - xlrd<2.0
deps/travis-37-locale.yaml:  - xlrd<2.0
```

I would say change one of the window and 38-slow to >2 leave the rest

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py]: Yea - agreed. I reworked the logic so whenever engine="xlrd" and xlrd is installed we either raise a ValueError or FutureWarning.

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py:1073]: Thanks - I missed this case. Should be fixed now.

File: pandas/io/excel/_base.py
Comment (rhshadrach) [reply pandas/io/excel/_base.py:1031]: If engine is not None, this is checked above; if it is None, it should only be possible to be an engine that is in the dictionary.  As such, I've added this back in as an assert.

**Comment by rhshadrach** [comment]
@cjw296 

> That code certainly works with bytes and urls,

Indeed, agreed. I had missed your usage of get_handle and concluded too quickly when I saw url tests failing.

**Comment by cjw296** [comment]
@rhshadrach - would you be interested in adding the warnings in the way I described in https://github.com/pandas-dev/pandas/pull/38522#issuecomment-749097931 and fixing up the remaining test failures there? The branch is open to commits from pandas people.

**Comment by rhshadrach** [comment]
@cjw296 No, I am not.

**Comment by cjw296** [comment]
> No, I am not.

*sigh*

File: pandas/io/excel/_base.py
Comment (jorisvandenbossche) [reply pandas/io/excel/_base.py:1031]: Before there was not an `else:` assigning an engine in any case (so it could end up being None), but with your changes that should indeed not be the case anymore.

**Comment by jreback** [comment]
@rhshadrach 

https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=50726&view=logs&j=404760ec-14d3-5d48-e580-13034792878f&t=f81e4cc8-d61a-5fb8-36be-36768e5c561a looks like a legit failure

**Comment by jorisvandenbossche** [comment]
@cjw296 adding the additional logic (fallback, warning, etc) to #38522 would also make it less clean. And in the meantime, effort in refining that logic has been done in this PR, so it would be harder to move that. Do you think there are still cases you covered in #38522 that are not covered here? As far as I can see, this PR uses much of the same inspection code, and also added the similar tests as you added in #38522.

**Review by jorisvandenbossche — COMMENTED** [review]
@rhshadrach thanks for the update! Additional changes look good

There is one case that also changed behaviour in this PR (in a good way), which raises on master, and that is reading from a xlsx file-like object without xlrd:

```
with open("test.xlsx", "rb") as f:
    df = pd.read_excel(f)
```

On master, this would raise an error if you have xlrd 2.0 installed (because it always uses xlrd for file-likes on master), but with this PR it will now correctly choose openpyxl (and raise a proper warning when using xlrd). 
Do you know if this improvement is covered by a test?

**Comment by rhshadrach** [comment]
> @rhshadrach
> 
> https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=50726&view=logs&j=404760ec-14d3-5d48-e580-13034792878f&t=f81e4cc8-d61a-5fb8-36be-36768e5c561a looks like a legit failure

@jreback - Agreed, it is, but I don't understand it. The warning is emitted when openpyxl is not installed but xlrd < 2.0 is and a non-xls file is used. Replicating this in my environment, the test passes with the warning emitted with the right stacklevel (checked manually). So maybe this is a Windows vs Linux issue?

What really doesn't make sense is that `actual_warning.filename` is `pandas\io\excel\_base.py` where the Warning is emitted, but the stacklevel is very clearly being set to either be 2 or 4. I don't see how it's possible that `_base.py` is then the filename.

Attempting to debug via CI now.

**Comment by jorisvandenbossche** [comment]
If it only fails on windows, I think it is also fine to add a `check_stacklevel=False` to the particular test ..

**Comment by rhshadrach** [comment]
> If it only fails on windows, I think it is also fine to add a `check_stacklevel=False` to the particular test ..

Will do, is this a known issue with Windows? Couldn't find an issue on github, will add one if it is.

**Comment by jorisvandenbossche** [comment]
Actually, it's a test asserting warnings about positional arguments, not about the engine. So that might interfere with the test, because it now also raises another warning than the one it is testing?

**Comment by rhshadrach** [comment]
@jorisvandenbossche Yes, that appears to me to be the reason why it's failing now. However, it passes locally for me and stepping through the assert code, I don't see any reason why it might not work on Windows. But most likely I'm missing some peculiarity about windows..

**Comment by jorisvandenbossche** [comment]
I think it is certainly fine to skip the check for stacklevel here (or mayble only skip it on windows, so it's still tested generally)

**Comment by jorisvandenbossche** [comment]
@rhshadrach sorry, I merged https://github.com/pandas-dev/pandas/pull/38456, which will give some merge conflict pains. I can also deal with it if you like

**Comment by rhshadrach** [comment]
@jorisvandenbossche - Not a problem. The merge was fairly straightforward, but your review would be appreciated to make sure I didn't mess any of it up. (I am also double checking it now)

The stacklevel issue was a bug in one of my previous PRs where the file path was hardcoded using '/'. Should be fixed now. Also your requested test has been added.

**Comment by jorisvandenbossche** [comment]
Doc changes look good after the merge!

File: pandas/tests/io/excel/test_readers.py
Comment (jorisvandenbossche) [inline pandas/tests/io/excel/test_readers.py:636]: Could also be left for a follow-up (time to get this merged ;-)), but this is not a super clear error message in case you accidentally pass a non-excel file to `read_excel`

We should probably check in the inspect function if it is a zip file (like is also done here: https://github.com/pandas-dev/pandas/pull/38522/files#diff-63200ddb7f5656b8ee868a28d9cb7720ffe50689b0e3fb0b4e15cc5c0ae80dd7R942), and if not raise an error like "file is not recognized as an excel file" or so.

**Comment by jreback** [comment]
@rhshadrach unfortunately a couple of legit looking failures. note happy to xfail those tests for those versions is fine.

**Comment by rhshadrach** [comment]
@jreback - the test suggested by @jorisvandenbossche detected that we were not defaulting to pyxlsb for xlsb files when engine is None; opened #38667 and xfailed the test.

**Comment by jreback** [comment]
> @jreback - the test suggested by @jorisvandenbossche detected that we were not defaulting to pyxlsb for xlsb files when engine is None; opened #38667 and xfailed the test.

excellent, ping on green-ish

**Comment by jreback** [comment]
thanks @rhshadrach amazing job here.

and thank you @cjw296 for your PR and the input.

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Comment by jorisvandenbossche** [comment]
Thanks @rhshadrach !

**Pull request merged**
